crates/runtime/src/instance/allocator/pooling/uffd.rs

```diff
@@ -0,0 +1,833 @@
+    max_instances: usize,
+    memories_start: usize,
+    memories_end: usize,
+    memory_size: usize,
+    max_memories: usize,
+    tables_start: usize,
+    tables_end: usize,
+    table_size: usize,
+    stacks_start: usize,
+    stacks_end: usize,
+    stack_size: usize,
+    page_size: usize,
+}
+
+impl AddressLocator {
+    fn new(instances: &InstancePool, stacks: &StackPool) -> Self {
+        let instances_start = instances.mapping.as_ptr() as usize;
+        let memories_start = instances.memories.mapping.as_ptr() as usize;
+        let memories_end = memories_start + instances.memories.mapping.len();
+        let tables_start = instances.tables.mapping.as_ptr() as usize;
+        let tables_end = tables_start + instances.tables.mapping.len();
+        let stacks_start = stacks.mapping.as_ptr() as usize;
+        let stacks_end = stacks_start + stacks.mapping.len();
+        let stack_size = stacks.stack_size;
+
+        // Should always have instances
+        debug_assert!(instances_start != 0);
+
+        Self {
+            instances_start,
+            instance_size: instances.instance_size,
+            max_instances: instances.max_instances,
+            memories_start,
+            memories_end,
+            memory_size: instances.memories.memory_size,
+            max_memories: instances.memories.max_memories,
+            tables_start,
+            tables_end,
+            table_size: instances.tables.table_size,
+            stacks_start,
+            stacks_end,
+            stack_size,
+            page_size: instances.tables.page_size,
+        }
+    }
+
+    /// This is super-duper unsafe as it is used from the handler thread
+    /// to access instance data without any locking primitives.
+    ///
+    /// It is assumed that the thread that owns the instance being accessed is
+    /// currently suspended waiting on a fault to be handled.
+    ///
+    /// Of course a stray faulting memory access from a thread that does not own
+    /// the instance might introduce a race, but this implementation considers
+    /// such to be a serious bug.
+    ///
+    /// If the assumption holds true, accessing the instance data from the handler thread
+    /// should, in theory, be safe.
+    unsafe fn get_instance(&self, index: usize) -> &Instance {
+        debug_assert!(index < self.max_instances);
+        &*((self.instances_start + (index * self.instance_size)) as *const Instance)
+    }
+
+    unsafe fn get_location(&self, addr: usize) -> Option<AddressLocation> {
+        // Check for a memory location
+        if addr >= self.memories_start && addr < self.memories_end {
+            let index = (addr - self.memories_start) / self.memory_size;
+            let memory_index = index % self.max_memories;
+            let memory_start = self.memories_start + (index * self.memory_size);
+            let page_index = (addr - memory_start) / WASM_PAGE_SIZE;
+            let instance = self.get_instance(index / self.max_memories);
```

Comment:
Hm I see now about my comment earlier having different index spaces for tables/memories will not work because of this. I can't really think of an alternative, so seems fine!

Reply:
I have some ideas around making this work so I'll see what I can do.

Reply:
I'm going to keep this current design rather than implementing free lists for the memory and table pools as I think it's desirable that the module limits guarantees `N` number of instances can be created, provided the modules being instantiated conform to the limits, rather than allow for memory or table pool exhaustion causing instantiation failures.

I think this is easier for users to reason about.  Resolving this conversation for now.